assets/js/shared/context/product-data-query.js

```diff
@@ -0,0 +1,53 @@
+/**
+ * External dependencies
+ */
+import { addQueryArgs } from '@wordpress/url';
+import apiFetch from '@wordpress/api-fetch';
+import { useEffect, useState, useRef } from '@wordpress/element';
+import { difference, concat } from 'lodash';
+
+/**
+ * Rendered by Product Data Context to query data for registered fields once children are rendered.
+ */
+const ProductDataQuery = ( { productId, fields, setProduct } ) => {
+	const [ productData, setProductData ] = useState( {} );
+	const retrievedFields = useRef( [] );
+
+	// If the product ID changes, reset fields.
+	useEffect( () => {
+		retrievedFields.current = [];
+		setProductData( {} );
+	}, [ productId ] );
+
+	// When product data changes, update the product context.
+	useEffect( () => {
+		setProduct( productData );
+	}, [ productData, setProduct ] );
+
+	// If the fields change, fetch new data.
+	useEffect( () => {
+		if ( ! productId || fields === undefined ) {
+			return;
+		}
+		const fieldsToFetch = difference( fields, retrievedFields.current );
+
+		if ( fieldsToFetch.length === 0 ) {
+			return;
+		}
+
+		apiFetch( {
+			path: addQueryArgs( `/wc/store/products/${ productId }`, {
+				_fields: [ ...fieldsToFetch, 'id' ],
```

Comment:
So this means that _at a minimum_ this will return just the id. I wonder if we should default to the entire product response if no fields are provided (`undefined`)?  Otherwise consuming code has to be aware of all the fields available on the endpoint if all fields are desired?

Reply:
This ProductDataQuery component has a pretty specific use-case, so I don't think it's unreasonable to make consuming code handle this, remembering the consumer could also pass in `[]` if desired. Atomic blocks are the only consumers.